From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nicolas Rochelemagne <rochelemagne@cpanel.net>
Date: Wed, 5 Jul 2017 12:29:41 -0600
Subject: [PATCH] BC Static shared memory for single malloc

do not update these files, use make regen_headers
- embed.h
- embedvar.h
- proto.h
---
 SOURCES/perl/embed.fnc         |  1 +
 SOURCES/perl/embed.h           |  1 +
 SOURCES/perl/embedvar.h        |  2 ++
 SOURCES/perl/intrpvar.h        |  8 ++++++++
 SOURCES/perl/perl.h            |  2 +-
 SOURCES/perl/proto.h           |  5 +++++
 SOURCES/perl/t/porting/regen.t |  2 ++
 SOURCES/perl/util.c            | 88 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 8 files changed, 108 insertions(+), 1 deletion(-)

diff --git a/SOURCES/perl/embed.fnc b/SOURCES/perl/embed.fnc
index 277bead9..a496fd1a 100644
--- a/SOURCES/perl/embed.fnc
+++ b/SOURCES/perl/embed.fnc
@@ -2715,6 +2715,7 @@ Ap	|int	|get_mstats	|NN perl_mstats_t *buf|int buflen|int level
 ATdpa	|Malloc_t|safesysmalloc	|MEM_SIZE nbytes
 ATdpa	|Malloc_t|safesyscalloc	|MEM_SIZE elements|MEM_SIZE size
 ATdpR	|Malloc_t|safesysrealloc|Malloc_t where|MEM_SIZE nbytes
+Anpa    |Malloc_t|bc_safesysrealloc|Malloc_t where|MEM_SIZE nbytes
 Anp     |void    |DeclareStaticMemory|Malloc_t from|Malloc_t to|MEM_SIZE size
 AdTp	|Free_t	|safesysfree	|Malloc_t where
 CrTp	|void	|croak_memory_wrap
diff --git a/SOURCES/perl/embed.h b/SOURCES/perl/embed.h
index ac70fc42..28bb2fac 100644
--- a/SOURCES/perl/embed.h
+++ b/SOURCES/perl/embed.h
@@ -72,6 +72,7 @@
 #define av_store_simple(a,b,c)	Perl_av_store_simple(aTHX_ a,b,c)
 #define av_undef(a)		Perl_av_undef(aTHX_ a)
 #define av_unshift(a,b)		Perl_av_unshift(aTHX_ a,b)
+#define bc_safesysrealloc	Perl_bc_safesysrealloc
 #define block_end(a,b)		Perl_block_end(aTHX_ a,b)
 #define block_gimme()		Perl_block_gimme(aTHX)
 #define block_start(a)		Perl_block_start(aTHX_ a)
diff --git a/SOURCES/perl/embedvar.h b/SOURCES/perl/embedvar.h
index c89e257d..1210c9d1 100644
--- a/SOURCES/perl/embedvar.h
+++ b/SOURCES/perl/embedvar.h
@@ -292,6 +292,8 @@
 #define PL_static_memory_buffer (vTHX->Istatic_memory_buffer)
 #define PL_static_memory_from   (vTHX->Istatic_memory_from)
 #define PL_static_memory_to     (vTHX->Istatic_memory_to)
+#define PL_static_shared_memory_position       (vTHX->Istatic_shared_memory_position)
+#define PL_static_shared_memory_table  (vTHX->Istatic_shared_memory_table)
 #define PL_statname		(vTHX->Istatname)
 #define PL_statusvalue		(vTHX->Istatusvalue)
 #define PL_statusvalue_posix	(vTHX->Istatusvalue_posix)
diff --git a/SOURCES/perl/intrpvar.h b/SOURCES/perl/intrpvar.h
index b9dac8fb..bf47cb73 100644
--- a/SOURCES/perl/intrpvar.h
+++ b/SOURCES/perl/intrpvar.h
@@ -1045,6 +1045,14 @@ PERLVARI(I, static_memory_from,        Malloc_t,       0)      /* first memory a
 PERLVARI(I, static_memory_to,  Malloc_t,       0)      /* last memory address where we have declared a usage of static memory */
 PERLVARI(I, static_memory_buffer, STATIC_MEMORY_AREA *, NULL)   /* buffer storage (linked list for now) */
 
+/* static_shared memory variable to use one single malloc at init time then share it (used by custom saferealloc) */
+PERLVARI(I, static_shared_memory_position, STATIC_MEMORY_AREA *, NULL) /* from and to position of the shared memory block
+                                                                            also note that size contains the number of entries in static_shared_memory_table
+                                                                        */
+PERLVARI(I, static_shared_memory_table, STATIC_MEMORY_AREA*, NULL) /* array of STATIC_MEMORY_AREA
+                                                                      to know what is the size hold by a pointer
+                                                                   */
+
 /* If you are adding a U8 or U16, check to see if there are 'Space' comments
  * above on where there are gaps which currently will be structure padding.  */
 
diff --git a/SOURCES/perl/perl.h b/SOURCES/perl/perl.h
index baf1719f..caf8f0af 100644
--- a/SOURCES/perl/perl.h
+++ b/SOURCES/perl/perl.h
@@ -1376,7 +1376,7 @@ Use L</UV> to declare variables of the maximum usable size on this platform.
 #else  /* MYMALLOC */
 #  define safemalloc  safesysmalloc
 #  define safecalloc  safesyscalloc
-#  define saferealloc safesysrealloc
+#  define saferealloc bc_safesysrealloc
 #  define safefree    safesysfree
 #  define CHECK_MALLOC_TOO_LATE_FOR(ch)		((void)0)
 #  define CHECK_MALLOC_TAINT(newval)		((void)0)
diff --git a/SOURCES/perl/proto.h b/SOURCES/perl/proto.h
index 12d1652c..9073b57e 100644
--- a/SOURCES/perl/proto.h
+++ b/SOURCES/perl/proto.h
@@ -35,6 +35,7 @@ PERL_STATIC_INLINE GV *	Perl_CvGV(pTHX_ CV *sv);
 #define PERL_ARGS_ASSERT_CVGV	\
 	assert(sv)
 #endif
+PERL_CALLCONV void  Perl_DeclareSharedMemoryUsage(Malloc_t where, MEM_SIZE size);
 PERL_CALLCONV void  Perl_DeclareStaticMemory(Malloc_t from, Malloc_t to, MEM_SIZE size);
 PERL_CALLCONV int	Perl_Gv_AMupdate(pTHX_ HV* stash, bool destructing);
 #define PERL_ARGS_ASSERT_GV_AMUPDATE	\
@@ -336,6 +337,10 @@ PERL_CALLCONV void	Perl_av_undef(pTHX_ AV *av);
 PERL_CALLCONV void	Perl_av_unshift(pTHX_ AV *av, SSize_t num);
 #define PERL_ARGS_ASSERT_AV_UNSHIFT	\
 	assert(av)
+PERL_CALLCONV Malloc_t	Perl_bc_safesysrealloc(Malloc_t where, MEM_SIZE nbytes)
+			__attribute__malloc__
+			__attribute__warn_unused_result__;
+
 PERL_CALLCONV OP*	Perl_bind_match(pTHX_ I32 type, OP *left, OP *right)
 			__attribute__warn_unused_result__;
 #define PERL_ARGS_ASSERT_BIND_MATCH	\
diff --git a/SOURCES/perl/t/porting/regen.t b/SOURCES/perl/t/porting/regen.t
index d684fdc1..3cf5d6e7 100644
--- a/SOURCES/perl/t/porting/regen.t
+++ b/SOURCES/perl/t/porting/regen.t
@@ -12,6 +12,8 @@ require './regen/regen_lib.pl';
 require './t/test.pl';
 $::NO_ENDING = $::NO_ENDING = 1;
 
+skip_all( "porting/regen.t disabled by cPanel" );
+
 if ( $^O eq "VMS" ) {
   skip_all( "- regen.pl needs porting." );
 }
diff --git a/SOURCES/perl/util.c b/SOURCES/perl/util.c
index d84ff81a..eea9afd0 100644
--- a/SOURCES/perl/util.c
+++ b/SOURCES/perl/util.c
@@ -213,6 +213,94 @@ Perl_safesysmalloc(MEM_SIZE size)
     return ptr;
 }
 
+/*
+	do a basic search in PL_static_shared_memory_table to guess
+	the size of the element hold by a pointer between
+	PL_static_shared_memory_position->from and PL_static_shared_memory_position->to
+	positions.
+
+	Note that the caller is responsible for checking that we are in this ranger earlier.
+	( view Perl_bc_safesysrealloc )
+
+    return value:
+        0: when cannot find the size malloc for that pointer
+        MEM_SIZE > 0: when you can find the size of the element pointing to
+
+*/
+MEM_SIZE
+_get_size_for_shared_memory_pointer( Malloc_t search ) {
+    int first, last, middle;
+
+    /* initialize our variables */
+    first = 0;
+    /* we stole the size here to store the number of entries in PL_static_shared_memory_table */
+    last  = (int) PL_static_shared_memory_position->size;
+
+    /* use a relative memory address to the first entry */
+    search = (Malloc_t) ((char *)search - (char *)PL_static_shared_memory_position->from);
+
+    /* while element not found
+        and we still have something to search (at least one entry in the array)
+    */
+    while( first <= last ) {
+            /* we always start looking from the middle */
+            middle = (first+last) / 2;
+            /* check if the current element in middle is the one we are looking foŕ */
+            if( PL_static_shared_memory_table[middle].from <= search
+                && search < PL_static_shared_memory_table[middle].to ) {
+#ifdef DEBUG_NEWXZ
+                /* extra safety / checks for debugging */
+                int delta = search - PL_static_shared_memory_table[middle].from;
+                int modulo = delta % PL_static_shared_memory_table[middle].size;
+                if ( modulo == 0 ) { /* need to point to the beginning of the struct */
+                    return PL_static_shared_memory_table[middle].size;
+                }
+                /* otherwise we are inside a pointer not malloced by us: there is something wrong ? */
+                return 0;
+#else
+                /* no strings there: just trust that we are pointing at the beginning of a struct */
+                return PL_static_shared_memory_table[middle].size;
+#endif
+            }
+            else
+            {  /* let's check which side of the table we should investigate */
+                if ( search < PL_static_shared_memory_table[middle].from )
+                    last = middle - 1; /* look on the left side */
+                else
+                    first = middle + 1; /* look on the right side */
+            }
+    }
+
+    return 0; /* size of 0 means we have not found it */
+}
+
+/* B::C wrapper around Perl_safesysrealloc */
+Malloc_t
+Perl_bc_safesysrealloc(Malloc_t where, MEM_SIZE size) {
+    /* check if we are in the range of the static shared memory pointers declared */
+	if ( LIKELY( !PL_static_shared_memory_table || !PL_static_shared_memory_position
+		|| where > PL_static_shared_memory_position->to /* check to first as it's most likely to be true */
+		|| where < PL_static_shared_memory_position->from
+		)
+	) {
+		return Perl_safesysrealloc(where, size);
+	} else {
+		MEM_SIZE oldsize = _get_size_for_shared_memory_pointer( where );
+		if ( oldsize ) {
+			Malloc_t newmem;
+			if ( size == 0 ) return NULL; /* nothing to free there return a NULL pointer */
+			Newxz(newmem, size, char); /* malloc a new memory pointer */
+            /* we do not want to do a free there ! the memory is static */
+			if ( oldsize > size ) oldsize = size; /* shrink memory (get the min) */
+			Copy(where, newmem, oldsize, char);
+			return newmem;
+		}
+	}
+
+	/* fallback to safesysrealloc */
+	return Perl_safesysrealloc(where, size);
+}
+
 /* used by B::C to declare static memory blocks */
 /*
 * sample usages:
